### PR TITLE
OPSEXP-2713: harden and refactor tomcat image build

### PR DESCRIPTION
**Pull request by alxgomz**
Ref: OPSEXP-2713

Downstream projects relying on tomcat base image would have to add a `USER root` directive before doing any `RUN` to avoid permissions issues.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -44,77 +42,111 @@ RUN \
...
 ARG BUILD_DIR=/build
 ARG INSTALL_DIR=/usr/local
-RUN set -eu; \
-  cd $BUILD_DIR; \
-  yum install -y gcc make openssl-devel expat-devel java-${JAVA_MAJOR}-openjdk-devel apr-devel redhat-rpm-config; \
-  cd tcnative/native; \
+COPY --from=tomcat_dist /build/tcnative $BUILD_DIR/tcnative
+WORKDIR ${BUILD_DIR}/tcnative/native
+SHELL ["/bin/bash", "-euo", "pipefail", "-c"]
+RUN yum install -y xmlstarlet gcc make openssl-devel expat-devel java-${JAVA_MAJOR}-openjdk-devel apr-devel redhat-rpm-config && yum clean all; \
```
Comment (gionn) [inline Dockerfile:91]:
```suggestion
RUN yum install -y gcc make openssl-devel expat-devel java-${JAVA_MAJOR}-openjdk-devel apr-devel redhat-rpm-config && yum clean all; \
```

File: Dockerfile [hunk 2]
```diff
@@ -44,77 +42,111 @@ RUN \
...
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t attr -n deployOnStartup -v false \
+  # Set RemoteIP valve for better monitoring/logging
+  -s '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t elem -n 'Valve' \
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/Valve[last()]' -t attr -n className -v org.apache.catalina.valves.RemoteIpValve \
+  # Do not leack server info within error pages
+  -s '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t elem -n 'Valve' \
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/Valve[last()]' -t attr -n className -v org.apache.catalina.valves.ErrorReportValve \
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/Valve[last()]' -t attr -n showServerInfo -v false \
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/Valve[last()]' -t attr -n showReport -v false \
+  # Do not leack runtime arguments and variables in logs
```
Comment (gionn) [inline Dockerfile:75]:
```suggestion
  # Do not leak runtime arguments and variables in logs
```

File: Dockerfile [hunk 3]
```diff
@@ -44,77 +42,111 @@ RUN \
...
+  -u '/Server/Service[@name="Catalina"]/Connector[@port=8080]/@Server' -v "" \
+  # Disable auto deployment of webapps
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t attr -n deployXML -v false \
+  -u '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/@autoDeploy' -v false \
+  -u '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/@unpackWARs' -v false \
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t attr -n deployOnStartup -v false \
+  # Set RemoteIP valve for better monitoring/logging
+  -s '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t elem -n 'Valve' \
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/Valve[last()]' -t attr -n className -v org.apache.catalina.valves.RemoteIpValve \
+  # Do not leack server info within error pages
```
Comment (gionn) [inline Dockerfile:70]:
```suggestion
  # Do not leak server info within error pages
```

File: Dockerfile [hunk 4]
```diff
@@ -44,77 +42,111 @@ RUN \
...
+  # Disable shutdown port
+  -d '/Server/@shutdown' -u '/Server/@port' -v -1 \
+  # Remove server banner
+  -u '/Server/Service[@name="Catalina"]/Connector[@port=8080]/@Server' -v "" \
+  # Disable auto deployment of webapps
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t attr -n deployXML -v false \
+  -u '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/@autoDeploy' -v false \
+  -u '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]/@unpackWARs' -v false \
+  -a '/Server/Service[@name="Catalina"]/Engine[@name="Catalina"]/Host[@name="localhost"]' -t attr -n deployOnStartup -v false \
+  # Set RemoteIP valve for better monitoring/logging
```
Comment (gionn) [inline Dockerfile:67]:
```suggestion
  # Enable RemoteIP valve for better monitoring/logging
```

File: Dockerfile [hunk 5]
```diff
@@ -44,77 +42,111 @@ RUN \
   gpg --batch --verify tomcat.tar.gz.asc tomcat.tar.gz && \
   tar -zxf tomcat.tar.gz -C /build/tomcat --strip-components=1 && \
   tar -zxf tcnative.tar.gz -C /build/tcnative --strip-components=1
+WORKDIR /build/tomcat
+# sh removes env vars it doesn't support (ones with periods)
+# https://github.com/docker-library/tomcat/issues/77
+RUN find ./bin/ -name '*.sh' -exec sed -ri 's|^#!/bin/sh$|#!/usr/bin/env bash|' '{}' +
+# Security improvements:
+# Remove server banner, Turn off loggin by the VersionLoggerListener, enable remoteIP valve so we know who we're talking to
```
Comment (gionn) [inline Dockerfile:50]:
```suggestion
```

they are enumerated later, maybe we can still add a mention in the readme

File: README.md [hunk 6]
```diff
@@ -35,6 +35,115 @@ The pinning approach provided in
...
+
+> We recommend that configuration applied in this image is generic and not
+> related to a single web application. A configuration change related to a
+> specific web application should be done in the web application image.
+
+### Using a different base distribution/image
+
+This project lets you use a different base distribution/image to create your
+Alfresco Tomcat base image.
+To do that you need to create additionnal, distro specific, stages in the
```
Comment (gionn) [inline README.md:96]:
```suggestion
To do that you need to create additional, distro specific, stages in the
```

File: Dockerfile (hunk 5 above)
Comment (alxgomz) [reply Dockerfile:50]: I'll just remove that. I don't want to introduce additional maintenance to keep the Dockerfile stage with the readme in sync.

**Review by gionn — APPROVED** [review]
worst case downstream images should drop their USER directives to avoid doing the same thing twice and/or temporarily override USER to root if they need to copy files over existing tomcat folders

**Review by pmacius — COMMENTED** [review]
Zrzut ekranu 2024-07-15 o 17 13 40
Change the color of the text. It looks like this on my side.

**Review by pmacius — APPROVED** [review]

**Review by pzhyland — APPROVED** [review]

**Review by dsibilio — APPROVED** [review]

**Review by gionn — APPROVED** [review]

**Pull request merged**